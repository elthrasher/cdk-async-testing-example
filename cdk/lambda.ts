import { Architecture, Function as LambdaFunction, Runtime } from '@aws-cdk/aws-lambda';
import { NodejsFunction } from '@aws-cdk/aws-lambda-nodejs';
import { Duration, Stack } from '@aws-cdk/core';

const functionNames = [
  'collect',
  'collectionFailure',
  'collectionSuccess',
  'openCollection',
  'paymentFailure',
  'paymentSuccess',
] as const;

export type lambdaFunctions = {
  [key in typeof functionNames[number]]: LambdaFunction;
};

const lambdaProps = {
  architectures: [Architecture.ARM_64],
  bundling: {
    externalModules: [],
  },
  runtime: Runtime.NODEJS_14_X,
  timeout: Duration.minutes(1),
};

export const getFunctions = (scope: Stack): lambdaFunctions => {
  return functionNames.reduce(
    (prev, fn) => ({
      ...prev,
      [fn]: new NodejsFunction(scope, `${fn}Function`, {
        ...lambdaProps,
        entry: `${__dirname}/../fns/${fn}.ts`,
        functionName: fn,
      }),
    }),
    {} as lambdaFunctions,
  );
};
