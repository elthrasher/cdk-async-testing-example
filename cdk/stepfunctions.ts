import { Chain, Choice, Condition, StateMachine } from '@aws-cdk/aws-stepfunctions';
import { LambdaInvoke } from '@aws-cdk/aws-stepfunctions-tasks';
import { Stack } from '@aws-cdk/core';

import { lambdaFunctions } from './lambda';

export const getStateMachine = (stack: Stack, fns: lambdaFunctions): StateMachine => {
  const openCollection = new LambdaInvoke(stack, 'OpenCollection', { lambdaFunction: fns.openCollection });
  const collect = new LambdaInvoke(stack, 'Collect', { lambdaFunction: fns.collect });
  const collectionFailure = new LambdaInvoke(stack, 'CollectionFailure', { lambdaFunction: fns.collectionFailure });
  const collectionSuccess = new LambdaInvoke(stack, 'CollectionSuccess', { lambdaFunction: fns.collectionSuccess });

  const isSuccessful = new Choice(stack, 'Is Collection Successful?');

  const chain = Chain.start(openCollection)
    // .next(new Wait(stack, 'WaitAround', { time: WaitTime.duration(Duration.seconds(30)) }))
    .next(collect)
    .next(
      isSuccessful
        .when(Condition.numberEquals('$.Payload.Status', 0), collectionSuccess)
        .when(Condition.numberEquals('$.Payload.Status', 1), collectionFailure),
    );

  return new StateMachine(stack, 'StateMachine', {
    definition: chain,
    stateMachineName: 'CollectionsStateMachine',
  });
};
