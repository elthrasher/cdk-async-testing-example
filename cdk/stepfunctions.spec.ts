import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

import { getFunctions } from './lambda';
import { getStateMachine } from './stepfunctions';

describe('Entire Stack', () => {
  test('returns lambda functions', () => {
    const app = new App();
    const stack = new Stack(app, 'TestStack', { env: { account: '123456789', region: 'us-east-1' } });
    const fns = getFunctions(stack);
    getStateMachine(stack, fns);
    const template = Template.fromStack(stack);
    const cfn = template.toJSON();
    const resources = cfn.Resources;
    const matchObject: { Parameters: Record<string, unknown>; Resources: Record<string, unknown> } = {
      Parameters: expect.any(Object),
      Resources: {},
    };
    Object.keys(resources).forEach((res) => {
      switch (resources[res].Type) {
        case 'AWS::Lambda::Function':
          matchObject.Resources[res] = {
            Properties: { Code: expect.any(Object) },
          };
          break;
        default:
          break;
      }
    });

    expect(cfn).toMatchSnapshot(matchObject);
    template.resourceCountIs('AWS::StepFunctions::StateMachine', 1);
  });
});
