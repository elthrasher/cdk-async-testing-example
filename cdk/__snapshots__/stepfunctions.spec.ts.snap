// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entire Stack returns lambda functions 1`] = `
Object {
  "Parameters": Any<Object>,
  "Resources": Object {
    "StateMachine2E01A3A5": Object {
      "DependsOn": Array [
        "StateMachineRoleDefaultPolicyDF1E6607",
        "StateMachineRoleB840431D",
      ],
      "Properties": Object {
        "DefinitionString": Object {
          "Fn::Join": Array [
            "",
            Array [
              "{\\"StartAt\\":\\"OpenCollection\\",\\"States\\":{\\"OpenCollection\\":{\\"Next\\":\\"Collect\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "openCollectionFunctionAD7A9939",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"Collect\\":{\\"Next\\":\\"Is Collection Successful?\\",\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "collectFunctionF1DD4A0C",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"Is Collection Successful?\\":{\\"Type\\":\\"Choice\\",\\"Choices\\":[{\\"Variable\\":\\"$.Payload.Status\\",\\"NumericEquals\\":0,\\"Next\\":\\"CollectionSuccess\\"},{\\"Variable\\":\\"$.Payload.Status\\",\\"NumericEquals\\":1,\\"Next\\":\\"CollectionFailure\\"}]},\\"CollectionSuccess\\":{\\"End\\":true,\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "collectionSuccessFunction5C4E05C2",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}},\\"CollectionFailure\\":{\\"End\\":true,\\"Retry\\":[{\\"ErrorEquals\\":[\\"Lambda.ServiceException\\",\\"Lambda.AWSLambdaException\\",\\"Lambda.SdkClientException\\"],\\"IntervalSeconds\\":2,\\"MaxAttempts\\":6,\\"BackoffRate\\":2}],\\"Type\\":\\"Task\\",\\"Resource\\":\\"arn:",
              Object {
                "Ref": "AWS::Partition",
              },
              ":states:::lambda:invoke\\",\\"Parameters\\":{\\"FunctionName\\":\\"",
              Object {
                "Fn::GetAtt": Array [
                  "collectionFailureFunction682979FB",
                  "Arn",
                ],
              },
              "\\",\\"Payload.$\\":\\"$\\"}}}}",
            ],
          ],
        },
        "RoleArn": Object {
          "Fn::GetAtt": Array [
            "StateMachineRoleB840431D",
            "Arn",
          ],
        },
        "StateMachineName": "CollectionsStateMachine",
        "TracingConfiguration": Object {
          "Enabled": true,
        },
      },
      "Type": "AWS::StepFunctions::StateMachine",
    },
    "StateMachineRoleB840431D": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "states.us-east-1.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "StateMachineRoleDefaultPolicyDF1E6607": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
                "xray:GetSamplingRules",
                "xray:GetSamplingTargets",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "openCollectionFunctionAD7A9939",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "collectFunctionF1DD4A0C",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "collectionSuccessFunction5C4E05C2",
                  "Arn",
                ],
              },
            },
            Object {
              "Action": "lambda:InvokeFunction",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::GetAtt": Array [
                  "collectionFailureFunction682979FB",
                  "Arn",
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "StateMachineRoleDefaultPolicyDF1E6607",
        "Roles": Array [
          Object {
            "Ref": "StateMachineRoleB840431D",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "collectFunctionF1DD4A0C": Object {
      "DependsOn": Array [
        "collectFunctionServiceRoleDefaultPolicyC2ED4D54",
        "collectFunctionServiceRoleBD487856",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "collect",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "collectFunctionServiceRoleBD487856",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "collectFunctionServiceRoleBD487856": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "collectFunctionServiceRoleDefaultPolicyC2ED4D54": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "collectFunctionServiceRoleDefaultPolicyC2ED4D54",
        "Roles": Array [
          Object {
            "Ref": "collectFunctionServiceRoleBD487856",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "collectionFailureFunction682979FB": Object {
      "DependsOn": Array [
        "collectionFailureFunctionServiceRoleDefaultPolicyB9F37612",
        "collectionFailureFunctionServiceRole662CE2FF",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "collectionFailure",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "collectionFailureFunctionServiceRole662CE2FF",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "collectionFailureFunctionServiceRole662CE2FF": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "collectionFailureFunctionServiceRoleDefaultPolicyB9F37612": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "collectionFailureFunctionServiceRoleDefaultPolicyB9F37612",
        "Roles": Array [
          Object {
            "Ref": "collectionFailureFunctionServiceRole662CE2FF",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "collectionSuccessFunction5C4E05C2": Object {
      "DependsOn": Array [
        "collectionSuccessFunctionServiceRoleDefaultPolicyA0766DBC",
        "collectionSuccessFunctionServiceRole727C0164",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "collectionSuccess",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "collectionSuccessFunctionServiceRole727C0164",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "collectionSuccessFunctionServiceRole727C0164": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "collectionSuccessFunctionServiceRoleDefaultPolicyA0766DBC": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "collectionSuccessFunctionServiceRoleDefaultPolicyA0766DBC",
        "Roles": Array [
          Object {
            "Ref": "collectionSuccessFunctionServiceRole727C0164",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "openCollectionFunctionAD7A9939": Object {
      "DependsOn": Array [
        "openCollectionFunctionServiceRoleDefaultPolicy7EB84761",
        "openCollectionFunctionServiceRole5D634737",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "openCollection",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "openCollectionFunctionServiceRole5D634737",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "openCollectionFunctionServiceRole5D634737": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "openCollectionFunctionServiceRoleDefaultPolicy7EB84761": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "openCollectionFunctionServiceRoleDefaultPolicy7EB84761",
        "Roles": Array [
          Object {
            "Ref": "openCollectionFunctionServiceRole5D634737",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "paymentFailureFunction2E4A54C2": Object {
      "DependsOn": Array [
        "paymentFailureFunctionServiceRoleDefaultPolicy78A9826D",
        "paymentFailureFunctionServiceRole7EB3B3AA",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "paymentFailure",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "paymentFailureFunctionServiceRole7EB3B3AA",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "paymentFailureFunctionServiceRole7EB3B3AA": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "paymentFailureFunctionServiceRoleDefaultPolicy78A9826D": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "paymentFailureFunctionServiceRoleDefaultPolicy78A9826D",
        "Roles": Array [
          Object {
            "Ref": "paymentFailureFunctionServiceRole7EB3B3AA",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "paymentSuccessFunction82B44889": Object {
      "DependsOn": Array [
        "paymentSuccessFunctionServiceRoleDefaultPolicyBD68F7D7",
        "paymentSuccessFunctionServiceRole417E5467",
      ],
      "Properties": Object {
        "Architectures": Array [
          "arm64",
        ],
        "Code": Any<Object>,
        "Environment": Object {
          "Variables": Object {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
            "NODE_OPTIONS": "--enable-source-maps",
          },
        },
        "FunctionName": "paymentSuccess",
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "paymentSuccessFunctionServiceRole417E5467",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 60,
        "TracingConfig": Object {
          "Mode": "Active",
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "paymentSuccessFunctionServiceRole417E5467": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "paymentSuccessFunctionServiceRoleDefaultPolicyBD68F7D7": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "xray:PutTraceSegments",
                "xray:PutTelemetryRecords",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "paymentSuccessFunctionServiceRoleDefaultPolicyBD68F7D7",
        "Roles": Array [
          Object {
            "Ref": "paymentSuccessFunctionServiceRole417E5467",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": Object {
    "CheckBootstrapVersion": Object {
      "Assertions": Array [
        Object {
          "Assert": Object {
            "Fn::Not": Array [
              Object {
                "Fn::Contains": Array [
                  Array [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  Object {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
